libs/eavmlib/src/spi.erl

```diff
@@ -38,36 +19,119 @@
 % SPDX-License-Identifier: Apache-2.0 OR LGPL-2.1-or-later
 %
 
+%%
+%% @doc
+%% This module provides an interface into the
+%% <a href="https://en.wikipedia.org/wiki/Serial_Peripheral_Interface">Serial Peripheral Interface</a> (SPI)
+%% supported on many devices.
+%%
+%% This module currently support the SPI "leader" (historically known as the "master") interface, allowing
+%% the leader to connect to one or more "follower" (historically known as "slave") devices.
+%%
+%% Users interact with this interface by creating an instance of the driver via the `open/1' function,
+%% with returns an opaque reference to the driver instance.  The `open/1' function takes a complex map
+%% structure, which configures the driver to connect to follower devices.  See the `open/1' documentation
+%% for details about the structure of this configuration map.
+%%
+%% Subsequent read and write operations use the SPI instance returned from the `open/1' function.  Users
+%% may read from a specific follower device at a specific address, write to the device at an address,
+%% or simultaneously read from and write to the device in a single transaction.
+%% @end
+%%
 -module(spi).
--export([open/1, read_at/3, write_at/4]).
 
+-export([open/1, read_at/4, write_at/5]).
+
+-type spi_peripheral() :: hspi | vspi.
 -type bus_config() :: [
-    {miso_io_num, non_neg_integer()}
-    | {mosi_io_num, non_neg_integer()}
-    | {sclk_io_num, non_neg_integer()}
+    {miso_io_num, non_neg_integer()} |
+    {mosi_io_num, non_neg_integer()} |
+    {sclk_io_num, non_neg_integer()} |
+    {spi_peripheral, spi_peripheral()}
 ].
 -type device_config() :: [
-    {clock_speed_hz, non_neg_integer()}
-    | {mode, non_neg_integer()}
-    | {spi_cs_io_num, non_neg_integer()}
-    | {address_len_bits, non_neg_integer()}
+    {clock_speed_hz, non_neg_integer()} |
+    {mode, 0..3} |
+    {spi_cs_io_num, non_neg_integer()} |
+    {address_len_bits, non_neg_integer()}
+].
+-type device_name() :: atom().
+-type params() :: [
+    {bus_config, bus_config()} |
+    {device_config, [{device_name(), device_config()}]}
 ].
--type param() :: {bus_config, bus_config()} | {device_config, device_config()}.
--type params() :: [param()].
+
 -type spi() :: pid().
 -type address() :: non_neg_integer().
 
+-define(DEFAULT_SPI_CLOCK_HZ, 1000000).
```

Comment:
I'm not really sure that having a default frequency rather having an explicit one is a good thing.